kafkaesk/retry.py

```diff
@@ -0,0 +1,354 @@
+
+        self._handlers = retry_handlers or {}
+        self._handler_cache: Dict[Type[Exception], Tuple[str, RetryHandler]] = {}
+
+        self._ready = False
+
+        if "RetryMessage:1" not in self.app.schemas:
+            self.app.schema("RetryMessage", version=1)(RetryMessage)
+
+    async def add_retry_handler(self, exception: Type[Exception], handler: "RetryHandler") -> None:
+        if exception in self._handlers:
+            raise ValueError(f"{exception} retry handler is already set")
+
+        self._handlers[exception] = handler
+        self.logger.debug(
+            f"Retry policy added new handler {handler.__class__.__name__} for {exception}"
+        )
+
+        if self._ready is True:
+            await self._handlers[exception].initialize(self)
+            self.logger.debug(
+                f"Retry policy initialized retry handler {handler.__class__.__name__}"
+            )
+
+        # Clear handler cache when handler is added
+        self._handler_cache = {}
+
+    async def remove_retry_handler(self, exception: Type[Exception]) -> None:
+        if exception in self._handlers:
+            handler = self._handlers[exception]
+
+            del self._handlers[exception]
+            self.logger.debug(
+                f"Retry policy removed handler {handler.__class__.__name__} for {exception}"
+            )
+
+            # Clear handler cache when handler is removed
+            self._handler_cache = {}
+
+            await handler.finalize()
+            self.logger.debug(f"Retry policy finalized retry handler {handler.__class__.__name__}")
+
+    async def initialize(self) -> None:
+        self.logger.debug("Retry policy initializing retry handlers...")
+        await asyncio.gather(*[handler.initialize(self) for handler in self._handlers.values()])
+        self.logger.debug("Retry policy initialized")
+        self._ready = True
+
+    async def finalize(self) -> None:
+        self._ready = False
+        self.logger.debug("Retry policy finalizing retry handlers...")
+        await asyncio.gather(*[handler.finalize() for handler in self._handlers.values()])
+        self.logger.debug("Retry policy finalized")
+
+    async def __call__(
+        self,
+        record: ConsumerRecord,
+        exception: Exception,
+        retry_history: Optional[RetryHistory] = None,
+    ) -> None:
+        if self._ready is not True:
+            raise RuntimeError("RetryPolicy is not initalized")
+
+        self.logger.debug(f"Handling msg retry: {record} {exception}")
+
+        with RETRY_POLICY_TIME.labels(
+            stream_id=record.topic, partition=record.partition, group_id=self.subscription.group,
+        ).time():
+            handler_key, handler = self._get_handler(exception)
+
+            if retry_history is None:
+                retry_history = RetryHistory()
+
+            # Add information about this failure to the history
+            retry_history.failures.append(
```

Comment:
Should we enforce a max size here?

something like

`retry_history.failures = retry_history.failures[-10:]`

?

Reply:
Done